portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -62,9 +63,21 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         }
         Gson gson = new Gson();
         response.setContentType("application/json;");
-        response.getWriter().println(gson.toJson(messages));
+        List<Message> limitedMessages = new ArrayList<>();
```

Comment:
You could reduce the nesting by doing:

if (messages.size() == 0) {
  response.getWriter().println(gson.toJson(limitedMessages));
  // Note that returning early helps you deal with the regular cases with less nesting.
  // Thus easier to spot bugs if any.
  return;
}
for (int i = 0; i < messages.size(); i++) {
  limitedMessages.add(messages.get(i));
}
response.getWriter().println(gson.toJson(limitedMessages));